lxd/api_vsock.go

```diff
@@ -0,0 +1,199 @@
+package main
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/lxc/lxd/lxd/instance"
+	"github.com/lxc/lxd/lxd/instance/instancetype"
+	"github.com/lxc/lxd/lxd/project"
+	"github.com/lxc/lxd/lxd/response"
+	"github.com/lxc/lxd/lxd/util"
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/version"
+)
+
+var vsockAPI10Cmd = APIEndpoint{
+	Get: APIEndpointAction{Handler: vsockAPI10Get, AllowUntrusted: true},
+}
+
+var vsockConfigCmd = APIEndpoint{
+	Path: "config",
+	Get:  APIEndpointAction{Handler: vsockConfigGet, AllowUntrusted: true},
+}
+
+var vsockMetaDataCmd = APIEndpoint{
+	Path: "meta-data",
+	Get:  APIEndpointAction{Handler: vsockMetaDataGet, AllowUntrusted: true},
+}
+
+var vsockDevicesCmd = APIEndpoint{
+	Path: "devices",
+	Get:  APIEndpointAction{Handler: vsockDevicesGet, AllowUntrusted: true},
+}
+
+func authenticateAgentCert(d *Daemon, r *http.Request) (bool, string, error) {
+	cacheUpdated := false
+	trusted := false
+	projectInstanceName := ""
+
+	if len(d.agentCerts) == 0 {
+		// Cache is empty -> load all instances and populate cache
+		err := updateAgentCertCache(d)
+		if err != nil {
+			return false, "", err
+		}
+
+		cacheUpdated = true
+	}
+
+	for _, cert := range r.TLS.PeerCertificates {
+		trusted, projectInstanceName = util.CheckTrustState(*cert, d.agentCerts, nil, false)
+		if trusted {
+			break
+		}
+	}
+
+	if !trusted {
+		// If the cache has been updated, but the certificate is not trusted, return Forbidden.
+		if cacheUpdated {
+			return false, "", nil
+		}
+
+		// Refresh the certificate cache as it might be out of date.
+		// Then try again.
+		err := updateAgentCertCache(d)
+		if err != nil {
+			return false, "", err
+		}
+
+		for _, cert := range r.TLS.PeerCertificates {
+			trusted, projectInstanceName = util.CheckTrustState(*cert, d.agentCerts, nil, false)
+			if trusted {
+				break
+			}
+		}
+
+		if !trusted {
+			return false, "", nil
+		}
+	}
+
+	return true, projectInstanceName, nil
+}
+
+func updateAgentCertCache(d *Daemon) error {
```

Comment:
I'm not sure that this is the best thing to do.

We can pull the remote address from the http.Request, this remote address can then be matched against the volatile key using a single DB query and then from that we can figure out the instance we need to pull the cert from.

That would save us from ever having to do this full listing of all instances.

Reply:
Caching may be fine but we'll need to be careful to invalidate when an instance stops as there's no guarantee of vsock id reuse for a particular instance.

I think if we get this down to a single `SELECT projects.name, instances.name FROM instances_config JOIN instances ON instances.id=instances_config.instance_id JOIN projects ON projects.id=instances.project_id WHERE instances.node_id=? AND instances_config.key='volatile.vsock_id' AND instances_config.value=?`, the DB cost will be minimal.

```
stgraber@dakara:~$ lxd sql global "SELECT projects.name, instances.name FROM instances_config JOIN instances ON instances.id=instances_config.instance_id JOIN projects ON projects.id=instances.project_id WHERE instances.node_id=1 AND instances_config.key='volatile.vsock_id' AND instances_config.value='32'"
+---------+-------+
|  name   | name  |
+---------+-------+
| default | win11 |
+---------+-------+
```

Reply:
I wonder if we maintain and clear the cache in the same place we maintain the agentReady boolean, that way we can clear it when it  disconnects.